vpr/src/route/lookahead_profiler.cpp

```diff
@@ -0,0 +1,173 @@
+    const auto& rr_graph = device_ctx.rr_graph;
+    auto& route_ctx = g_vpr_ctx.routing();
+
+    // If csv file hasn't been opened, open it and write out column headers
+    if (is_empty) {
+        lookahead_verifier_csv.open("lookahead_verifier_info.csv", std::ios::out);
+
+        if (!lookahead_verifier_csv.is_open()) {
+            VTR_LOG_ERROR("Could not open lookahead_verifier_info.csv", "error");
+        }
+
+        lookahead_verifier_csv
+            << "iteration no."
+            << ",source node"
+            << ",sink node"
+            << ",sink block name"
+            << ",sink atom block model"
+            << ",sink cluster block type"
+            << ",sink cluster tile height"
+            << ",sink cluster tile width"
+            << ",current node"
+            << ",node type"
+            << ",node length"
+            << ",num. nodes from sink"
+            << ",delta x"
+            << ",delta y"
+            << ",actual cost"
+            << ",actual delay"
+            << ",actual congestion"
+            << ",predicted cost"
+            << ",predicted delay"
+            << ",predicted congestion"
+            << ",criticality"
+            << std::endl;
+
+        is_empty = false;
+    }
+
+    if (!lookahead_verifier_csv.is_open())
+        return;
+
+    // The default value in RouteTree::update_from_heap() is -1; only calls which descend from route_net()
+    // pass in an iteration value, which is the only context in which we want to profile.
+    if (iteration < 1)
+        return;
+
+    RRNodeId source_inode = branch_inodes.back(); // Not necessarily an actual SOURCE node.
+    RRNodeId sink_inode = branch_inodes.front();
+
+    VTR_ASSERT(rr_graph.node_type(sink_inode) == SINK);
+
+    /* Get sink node attributes (atom block name, atom block model, cluster type, tile dimensions) */
+
+    if (atom_block_names.find(sink_inode) == atom_block_names.end()) {
+        if (net_id != ParentNetId::INVALID() && target_net_pin_index != OPEN) {
+            atom_block_names[sink_inode] = net_list.block_name(net_list.net_pin_block(net_id, target_net_pin_index));
+
+            AtomBlockId atom_block_id = g_vpr_ctx.atom().nlist.find_block(atom_block_names[sink_inode]);
+            atom_block_models[sink_inode] = g_vpr_ctx.atom().nlist.block_model(atom_block_id)->name;
+
+            ClusterBlockId cluster_block_id = atom_to_cluster(atom_block_id);
+
+            cluster_block_types[sink_inode] = g_vpr_ctx.clustering().clb_nlist.block_type(cluster_block_id)->name;
+
+            auto tile_type = physical_tile_type(cluster_block_id);
+            tile_dimensions[sink_inode] = std::pair(std::to_string(tile_type->width), std::to_string(tile_type->height));
+        } else {
+            atom_block_names[sink_inode] = "--";
+            atom_block_models[sink_inode] = "--";
+            cluster_block_types[sink_inode] = "--";
+            tile_dimensions[sink_inode] = {"--", "--"};
+        }
+    }
```

Comment:
Do you mean that I should initialize them as empty maps in the constructor?

Reply:
What I had in mind was initializing them with all sinks ids as key and proper values. However, this can't be done when LookaheadProfiler has a global instance.

Reply:
I think storing multiple std::string values for each sink id is not memory efficient. For `atom_block_models`, `cluster_block_types`, `and tile_dimensions`, the number of unique values is very low. Even for `atom_block_names` where each unique value appears only a few times, you don't need to store multiple copies of each block name.

std::string_view, vtr::interned_string can be used to avoid unnecessary string copies.
It is also possible to store atom and cluster block ids and use global netlist objects to retrieve references to block/model names and tile sizes.

Reply:
I decided to change these variables to store IDs and type pointers. Dereferencing the pointers may be less runtime-efficient, but to use ``std::string_view``, I think I'd have to create a separate ``std::set`` which contains all the ``string_view``s encountered so far to see if we need to fetch the info for the current node, which seems less clean.